agent/src/main/java/com/microsoft/applicationinsights/agent/internal/MainEntryPoint.java

```diff
@@ -46,21 +49,144 @@ public static void premain(Instrumentation instrumentation, File agentJarFile) {
+            // TODO this has consequences if app is using AI SDK
+            return;
+        }
+
+        Global.isW3CEnabled = builtInConfiguration.isW3cEnabled();
+        Global.isW3CBackportEnabled = builtInConfiguration.isW3CBackportEnabled();
+
+        List<InstrumentationDescriptor> instrumentationDescriptors = getInstrumentationDescriptors(agentConfiguration);
 
         ConfigServiceFactory configServiceFactory = new SimpleConfigServiceFactory(instrumentationDescriptors,
-                getInstrumentationConfig());
+                getInstrumentationConfig(builtInConfiguration));
 
         EngineModule.createWithSomeDefaults(instrumentation, tmpDir, Global.getThreadContextThreadLocal(),
                 instrumentationDescriptors, configServiceFactory, agent,
                 Collections.singletonList("com.microsoft.applicationinsights."), agentJarFile);
     }
 
-    private static Map<String, Map<String, Object>> getInstrumentationConfig() {
+    private static List<InstrumentationDescriptor> getInstrumentationDescriptors(AgentConfiguration agentConfiguration)
+            throws IOException {
+
+        AgentBuiltInConfiguration builtInConfiguration = agentConfiguration.getBuiltInConfiguration();
+        boolean httpEnabled = builtInConfiguration.isHttpEnabled();
+        boolean jdbcEnabled = builtInConfiguration.isJdbcEnabled();
+        boolean redisEnabled = builtInConfiguration.isRedisEnabled();
+
+        List<InstrumentationDescriptor> instrumentationDescriptors = new ArrayList<>();
+        for (InstrumentationDescriptor instrumentationDescriptor : InstrumentationDescriptors.read()) {
+            switch (instrumentationDescriptor.name()) {
+                case "apache-http-client":
+                case "okhttp":
+                    if (httpEnabled) {
+                        instrumentationDescriptors.add(instrumentationDescriptor);
+                    }
+                    break;
+                case "jdbc":
+                    if (jdbcEnabled) {
+                        instrumentationDescriptors.add(instrumentationDescriptor);
+                    }
+                    break;
+                case "redis":
+                    if (redisEnabled) {
+                        instrumentationDescriptors.add(instrumentationDescriptor);
+                    }
+                    break;
+                default:
+                    instrumentationDescriptors.add(instrumentationDescriptor);
+                    break;
+            }
+        }
+
+        InstrumentationDescriptor instrumentationDescriptor = buildCustomInstrumentation(agentConfiguration);
+        if (instrumentationDescriptor != null) {
+            // need to test before enabling this
+            // instrumentationDescriptors.add(instrumentationDescriptor);
+        }
+        return instrumentationDescriptors;
+    }
+
+    private static InstrumentationDescriptor buildCustomInstrumentation(AgentConfiguration agentConfiguration) {
```

Comment:
Are there tests for this scenario? if not, let's create an issue/task

Reply:
This code is not called from anywhere yet (it's a possible-feature-in-progress)